### ISSUE TITLE
Improve OSSIndex Analyzer batching and resilience

### ISSUE THREAD

**Issue by syalioune** [DependencyTrack/dependency-track#2023]
### Current Behavior:

[OssIndexAnalysisTask](https://github.com/DependencyTrack/dependency-track/blob/master/src/main/java/org/dependencytrack/tasks/scanners/OssIndexAnalysisTask.java) paginate its request up to **100 purls** per calls. However, there are rooms for improvements :

- The allowed maximum number of purl per call is now [128](https://ossindex.sonatype.org/doc/rest) instead of 100
- Pruning component for which the component analysis cache is still valid [is not done before pagination](https://github.com/DependencyTrack/dependency-track/blob/master/src/main/java/org/dependencytrack/tasks/scanners/OssIndexAnalysisTask.java#L164), this can result in inefficient batching like in the example below where we could have done **7 requests instead of 16 requests** to OSSIndex (5 requests with 128 components per page)

```
2022-10-08 19:20:12,563 INFO [OssIndexAnalysisTask] Analyzing 47 component(s)
2022-10-08 19:20:17,576 INFO [OssIndexAnalysisTask] Analyzing 91 component(s)
2022-10-08 19:20:19,946 INFO [OssIndexAnalysisTask] Analyzing 74 component(s)
2022-10-08 19:20:27,405 INFO [OssIndexAnalysisTask] Analyzing 71 component(s)
2022-10-08 19:20:30,803 INFO [OssIndexAnalysisTask] Analyzing 48 component(s)
2022-10-08 19:20:34,584 INFO [OssIndexAnalysisTask] Analyzing 86 component(s)
2022-10-08 19:20:36,513 INFO [OssIndexAnalysisTask] Analyzing 13 component(s)
2022-10-08 19:20:38,307 INFO [OssIndexAnalysisTask] Analyzing 2 component(s)
2022-10-08 19:20:41,403 INFO [OssIndexAnalysisTask] Analyzing 2 component(s)
2022-10-08 19:20:45,016 INFO [OssIndexAnalysisTask] Analyzing 2 component(s)
2022-10-08 19:20:46,315 INFO [OssIndexAnalysisTask] Analyzing 1 component(s)
2022-10-08 19:20:57,674 INFO [OssIndexAnalysisTask] Analyzing 41 component(s)
2022-10-08 19:20:59,069 INFO [OssIndexAnalysisTask] Analyzing 46 component(s)
2022-10-08 19:21:04,412 INFO [OssIndexAnalysisTask] Analyzing 42 component(s)
2022-10-08 19:21:06,030 INFO [OssIndexAnalysisTask] Analyzing 41 component(s)
2022-10-08 19:21:07,461 INFO [OssIndexAnalysisTask] Analyzing 31 component(s)
```
- There are no retry (with exponential beckoff) or throttling mechanisms implemented. In case of any HTTP error (like 429 for rate threshold breach), only a log and notification are issued.

### Proposed Behavior:

Implement solutions for each of the three points mentioned above.

**Comment by nscuro** [comment]
Definitely something we can improve, thanks for reporting!

**Comment by msymons** [comment]
Does the OSSIndex batching remove duplicates?  ie, another opportunity for improving efficiency?

I have several score microservice projects that will all use pretty much the same components... so,  there would be no need to query (say) snakeyaml v1.27 multiple times in the space of a few seconds.  And so on an so forth for component after component.

**Comment by syalioune** [comment]
> Does the OSSIndex batching remove duplicates? ie, another opportunity for improving efficiency?

No, it does not ! Duplicate removal could however only be done per project as that is the scope of the list of components received by OSSIndex analyzer.

**Comment by valentijnscholten** [comment]
If I read the code correctly, analysis results are cached based on the purl. The cache is in the database, so works across projects.

**Comment by nscuro** [comment]
This topic is also related to #374. We need to decouple the analyzer logic in such a way that it's possible to scan components that are not tracked. This will require some refactoring.

> There are no retry (with exponential beckoff) or throttling mechanisms implemented. In case of any HTTP error (like 429 for rate threshold breach), only a log and notification are issued.

I'm eyeballing resilience4j for this as it has great retry support (https://resilience4j.readme.io/docs/retry), including exponential back-off strategies (https://resilience4j.readme.io/docs/retry#use-a-custom-intervalfunction). It also has circuit breaker and bulkhead support that may come in handy for other API clients we have (potentially #1861).

**Comment by syalioune** [comment]
> If I read the code correctly, analysis results are cached based on the purl. The cache is in the database, so works across projects.

That is true if the component analysis cache is still valid. However, if the cache is invalid (or not populated) and analysis is trigerred at the same time for N projects having the same component (i.e purl), there's a chance that N requests for the same purl would be performed before updating the cache.

**Comment by syalioune** [comment]
> I'm eyeballing resilience4j for this as it has great retry support (https://resilience4j.readme.io/docs/retry), including exponential back-off strategies (https://resilience4j.readme.io/docs/retry#use-a-custom-intervalfunction). It also has circuit breaker and bulkhead support that may come in handy for other API clients we have (potentially #1861).

I was also thinking about resilience4j 👍

**Comment by nscuro** [comment]
> However, if the cache is invalid (or not populated) and analysis is trigerred at the same time for N projects having the same component (i.e purl), there's a chance that N requests for the same purl would be performed before updating the cache.

Any idea how this part could be optimized, outside of excessive locking / synchronization? I saw myself in the [acceptance phase](https://en.m.wikipedia.org/wiki/Five_stages_of_grief) already, assuming that we just have to deal with the fact that we may do redundant calls sometimes.

**Comment by valentijnscholten** [comment]
I am wondering if this actually happens a lot? Wouldn't you nede quite a lot of things to align to have "lots of duplicate calls" happen? 

You would need:
- Lots of projects with the same component/version/purl
- Lots of these projects being uploaded "at the same time"
- Lots of the analysis events inside DT being executed "at the same time"

Sounds like this would only happen occasionally and might not be worth adding synchronization for? Synchronization might make things faster in the happy flow, but once you get errors/timeous/hiccups it can be painful to debug/solve.

**Comment by syalioune** [comment]
I think that synchronization would be an excessive and inefficient solution globally. Theoritically, I see two possible paths :

- Making the analyzer task a singleton operating based on an queue (in-memory or not) : No more problems because they would be only one thread working and serializing the calls. It's simple but not sure if it's relevant performance wise
- Using a combination of object pools and sharding based on queues (in-memory or not) : Having a pool of analyzer objects with the proper logic fetching components to process from a dedicated queue. Upstream process have to be updated to always send (or it can be wrapped) identical purls to the same queue (using some kind of hashed based partitioning). This way, there would be no need for synchronization between the analyzer objects as identitical purls would always be processed by the same analyzer

Anyway, the occurences of that situation should be assessed as @valentijnscholten said.

**Comment by nscuro** [comment]
Implemented in #2101.
